src/librustc_mir/transform/simplify.rs

```diff
@@ -382,6 +382,11 @@ impl<'a, 'tcx> Visitor<'tcx> for DeclMarker<'a, 'tcx> {
                             }
                         }
                     }
+                } else if let StatementKind::Assign(box (p, Rvalue::Discriminant(d))) = &stmt.kind {
```

Comment:
Are there any `Rvalue`s (except those that reference a `ConstKind::Unevaluated`, which I'm working on) that have side effects? I think we could use a visitor here looking for `ConstKind::Unevaluated` and remove the assignment if there is no `Unevaluated` in that `Rvalue`

Reply:
Hmm... I don't think we have side effects in the sense that there might be a user-defined function call or something hiding in an `Rvalue` (excluding `ConstKind::Unevaluated`). However there are somethings that might be termed side effects like:

- de-referencing a pointer
  - Removing a deref means we could potentially be removing UB if the pointer isn't actually valid. That *seems* ok to me. 

- Boxing values
  - I believe this lowers to a call to the global allocator which means you could observe this *not* happening via a custom allocator. 

- {Checked,}Binary ops could over/under flow.

In general, `Rvalues` except for these seem side-effect free to me (excluding `ConstKind::Unevaluated` and things that contain `Place` values with projections):

- `BinaryOp`
- `CheckedBinaryOp`
- `Cast`
- `NullaryOp`
- `UnaryOp`

Reply:
`CheckedBinaryOp` doesn't have a side-effect as over/under flow will just return that it had over/under flowed. `UnaryOp` is used to implement both negation and bitwise not. The later is always side-effect free. `Cast(CatKind::Pointer(_), _, _)` is side-effect free I think, but `Cast(CastKind::Misc, _, _)` could remove a ptr-int cast or int-ptr cast. `NullaryOp(NullOp::SizeOf, ty)` only needs to ensure that `layout_of` doesn't error on the type.

> Boxing values
> 
>     * I believe this lowers to a call to the global allocator which means you could observe this _not_ happening via a custom allocator.

Yes, it lowers to the `exchange_malloc` lang item, which uses `Global.alloc`:

https://github.com/rust-lang/rust/blob/dd155df0a69338757ca39a2a606a6accb7b8d342/src/liballoc/alloc.rs#L225

Removing this allocation would be equivalent to any function in liballoc or libstd changing when and how much it allocates though, so I don't think anybody can depend on it anyway.